### PR TITLE
Improve code coverage for Core Extensions (File & Book)

### LINKED ISSUE #4715: Improve code coverage for Core Extensions (File & Book)
**Particular area and files to be tested**
* `core/src/main/java/org/kiwix/kiwixmobile/core/extensions/FileExtensions.kt`
* `core/src/main/java/org/kiwix/kiwixmobile/core/extensions/BookExtensions.kt`

**Current Status of the tests**
These extension files currently have low coverage. They contain critical utility functions for file operations and book metadata extraction that need verification.

**Particular Behaviour to be tested**

**1. FileExtensions.kt**
* **File Operations:** Verify `isFileExist`, `deleteFile`, and `canReadFile` correctly wrap standard file operations in the IO dispatcher.
* **Space Calculation:** Verify `freeSpace` and `totalSpace` return valid values.
* **Coroutines:** Ensure these suspend functions execute safely without blocking the main thread.

**2. BookExtensions.kt**
* **Favicon Extraction:** Verify `getFavicon()` correctly extracts and encodes the illustration/favicon from a `Book` object.
* **Base64 Encoding:** Verify that raw image data is correctly converted to a Base64 string.
* **Error Handling:** Verify that `runCatching` blocks correctly handle nulls or exceptions (like missing data) without crashing.

Image

Image

### PR DESCRIPTION
**Pull request by yush-1018**
Fixes #4715

Added unit tests for FileExtensions.kt and instrumentation tests for BookExtensions.kt to improve code coverage.

### CODE REVIEW

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz @yush-1018 
I reviewed the errors it is because of the the Issue i raised so i'll cover it and provide.

**Comment by yush-1018** [comment]
@MohitMaliDeveloper ,

Please review this PR once.

Thanks!

File: core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+  @Test
+  fun `isFileExist should return false for non-existing file`() = runTest(testDispatcher) {
+    val nonExistentFile = File("non_existent_file_path.tmp")
+    assertFalse(nonExistentFile.isFileExist(testDispatcher))
+  }
+
+  @Test
+  fun `deleteFile should delete the file and return true`() = runTest(testDispatcher) {
+    assertTrue(tempFile.exists(), "File should exist before deletion")
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt]: Instead of `tempFile.exists()` use `tempFile.isFileExist()`.

File: core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    val result = tempFile.deleteFile(testDispatcher)
+    assertTrue(result, "deleteFile should return true")
+    assertFalse(tempFile.exists(), "File should not exist after deletion")
+  }
+
+  @Test
+  fun `canReadFile should return true for readable file`() = runTest(testDispatcher) {
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt:74]: We should write a test to verify `canReadFile` returns false for those files that can not be read.

File: core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    assertTrue(tempFile.canReadFile(testDispatcher))
+  }
+
+  @Test
+  fun `freeSpace should return a value greater than zero`() = runTest(testDispatcher) {
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt]: Instead of testing greater than zero, we should test the actual freeSpace.

File: core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    val free = tempFile.freeSpace(testDispatcher)
+    assertTrue(free > 0, "Free space should be greater than zero")
+  }
+
+  @Test
+  fun `totalSpace should return a value greater than zero`() = runTest(testDispatcher) {
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt]: Same for this.

File: core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    val total = tempFile.totalSpace(testDispatcher)
+    assertTrue(total > 0, "Total space should be greater than zero")
+  }
+
+  @Test
+  fun `totalSpace should be greater than or equal to freeSpace`() = runTest(testDispatcher) {
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt:107]: In this, we should also test the real values.

Set the totalSpace and freeSpace to the file, and then test it.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/extensions/BookExtensionsTest.kt [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    val result = null.getFavicon()
+    assertNull(
+      "getFavicon on a null Book should return null",
+      result
+    )
+  }
+
+  @Test
+  fun getFaviconShouldReturnEmptyStringWhenIllustrationThrowsException() {
+    // Book() with no backing archive will throw when getIllustration is called,
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/extensions/BookExtensionsTest.kt]: We are writing this in `androidTest` to test with the actual libkiwix. So, we should test the real favicon as well(with a real ZIM file).

Also, we should test the real favicon, and if the favicon url is null, it returns the default favicon. We should test those scenarios as well.

**Comment by MohitMaliFtechiz** [comment]
Rebase your PR as well.

**Comment by harshsomankar123-tech** [comment]
@MohitMaliFtechiz PTAL. Since @yush-1018 is busy with other work, he shared his fork with me to continue his contribution. Thanks!

File: core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt [hunk 7]
```diff
@@ -69,22 +69,34 @@ class FileExtensionsTest {
   }
 
   @Test
-  fun `freeSpace should return a value greater than zero`() = runTest(testDispatcher) {
-    val free = tempFile.freeSpace(testDispatcher)
-    assertTrue(free > 0, "Free space should be greater than zero")
+  fun `canReadFile should return false for non-readable file`() = runTest(testDispatcher) {
+    tempFile.setReadable(false)
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt]: This test is OS dependent. See it is failing on my machine. We should not write the OS dependent test cases.

Image

**Comment by MohitMaliFtechiz** [comment]
@harshsomankar123-tech, @yush-1018, There are some small improvements needed. I have made them. Once CI is green. We will merge this.

File: core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt (hunk 7 above)
Comment (harshsomankar123-tech) [reply core/src/test/java/org/kiwix/kiwixmobile/core/extensions/FileExtensionsTest.kt]: Screenshot 2026-03-23 at 12 34 45Yeah u are right This is likely OS-dependent because it passes on my local setup.
